### DISCUSSION TITLE
What was the error in my collision with mesh?

### DISCUSSION THREAD

**Discussion by bluejin404** [google-deepmind/mujoco#1912] [Asking for Help] [answered]
I'm new to mujoco and trying to write a MJCF from scratch, but the collision isn't working.

This is my robot in mujoco.
![image](https://github.com/user-attachments/assets/283928f2-7c32-4a7e-b547-ef301c42903a)

I hit the H key to see the convex hulls of your meshes. but they are just overlapping each other without any disturbance.
![image](https://github.com/user-attachments/assets/5f138418-c324-4928-a130-6c9a93ba77e5)

Here's my MJCF. would you help me with fixing this? My ChatGPT and Claude didn't give me any help.
`<mujoco model="softmachine0">

  <asset>
    <mesh name="inner-femur" file="Inner Femur.obj" />
    <mesh name="tibia" file="Tibia.obj" />
  </asset>

  <worldbody>
    <light diffuse=".5 .5 .5" pos="0 0 3" dir="0 0 -1" />
    <geom type="plane" size="1 1 0.1" rgba="0.5 0.5 0.5 1" />

    <body pos="0 0 0.3">
      <geom type="cylinder" size="0.0275" fromto="0 0 0 0.041 0 0" mass="0.31" rgba="0.2 0.2 0.2 1" />
      <joint type="free" />

      <body name="femur">
        <joint name="hip-pitch" type="hinge" axis="1 0 0" />
        <geom type="mesh" mesh="inner-femur" mass="0.424" contype="1" conaffinity="1" />

        <body>
          <geom type="cylinder" size="0.0275" fromto="0 0 -0.06 0.041 0 -0.06" mass="0.31"
            rgba="0.2 0.2 0.2 1" />
        </body>

        <body>
          <geom type="cylinder" size="0.0265"
            fromto="0 0.015529 -0.117956 0.031 0.015529 -0.117956" mass="0.217" rgba="0.2 0.2 0.2 1" />
        </body>

        <body name="tibia" pos="-0.006 0.015529 -0.117956">
          <joint name="knee" type="hinge" axis="1 0 0" />
          <geom type="mesh" mesh="tibia" mass="0.148" contype="1" conaffinity="1" />
        </body>
      </body>
    </body>
  </worldbody>

  <actuator>
    <motor name="hip-pitch-motor" joint="hip-pitch" ctrllimited="true" ctrlrange="-0.05 0.05"
      gear="36"
      user="90" />
    <motor name="knee-motor" joint="knee" ctrllimited="true" ctrlrange="-0.05 0.05" gear="10"
      user="90" />
  </actuator>
</mujoco>`

**Comment by yuvaltassa** [comment]
What do you mean "isn't working"? Not clear what the issue is.

**Reply by bluejin404** [reply]
Two mesh geoms overlap smoothly. I think they need to bounce after a collision. I tried a similar experience with much simpler box mesh geoms and the collision simulation worked.

**Reply by yuvaltassa** [reply]
Can you give all your geoms names, repost the XML and then tell me which geoms are colliding that you think shouldn't collide (or aren't colliding that should collide)

**Reply by bluejin404** [reply]
Sorry, here's my full XML(I didn't noticed that code is cropped) "femur" and "tibia" do not seems to collide, but they should. I also attached the obj files(the zip file) of femur and tibia.
[mesh.zip](https://github.com/user-attachments/files/16675316/mesh.zip)

```
<mujoco model="softmachine0">

  <asset>
    <mesh name="inner-femur" file="Inner Femur.obj" />
    <mesh name="tibia" file="Tibia.obj" />
  </asset>

  <worldbody>
    <light diffuse=".5 .5 .5" pos="0 0 3" dir="0 0 -1" />
    <geom type="plane" size="1 1 0.1" rgba="0.5 0.5 0.5 1" />

    <body pos="0 0 0.3">
      <geom type="cylinder" size="0.0275" fromto="0 0 0 0.041 0 0" mass="0.31" rgba="0.2 0.2 0.2 1" />
      <joint type="free" />


      <body name="femur">
        <joint name="hip-pitch" type="hinge" axis="1 0 0" />
        <geom type="mesh" mesh="inner-femur" mass="0.424" />

        <body>
          <geom type="cylinder" size="0.0275" fromto="0 0 -0.06 0.041 0 -0.06" mass="0.31"
            rgba="0.2 0.2 0.2 1" />
        </body>

        <body>
          <geom type="cylinder" size="0.0265"
            fromto="0 0.015529 -0.117956 0.031 0.015529 -0.117956" mass="0.217" rgba="0.2 0.2 0.2 1" />
        </body>

        <body name="tibia" pos="-0.012 0.015529 -0.117956">
          <joint name="knee" type="hinge" axis="1 0 0" />
          <geom type="mesh" mesh="tibia" mass="0.148" />
        </body>
      </body>
    </body>
  </worldbody>

  <actuator>
    <motor name="hip-pitch-motor" joint="hip-pitch" ctrllimited="true" ctrlrange="-0.05 0.05"
      gear="36"
      user="90" />
    <motor name="knee-motor" joint="knee" ctrllimited="true" ctrlrange="-0.05 0.05" gear="10"
      user="90" />
  </actuator>
</mujoco>
```

**Reply by yuvaltassa** [reply, marked as answer]
Tibia is a child of femur so they don't collide. If you really want you can disable [parent-child collision filtering](https://mujoco.readthedocs.io/en/latest/XMLreference.html#option-flag-filterparent), but I doubt that will give you want you want, you'll have all sorts of unexpected collisions. Your best best is to just explicitly specify [two geoms](https://mujoco.readthedocs.io/en/latest/XMLreference.html#contact-pair) that you want to collide.
